File: airesources/Julia/hlt.jl
const STILL = 0
const NORTH = 1
const EAST = 2
const SOUTH = 3
const WEST = 4

const DIRECTIONS = [a for a in 0:5]
const CARDINALS = [a for a in 1:5]

const ATTACK = 0
const STOP_ATTACK = 1

type Location
  x :: Int64
  y :: Int64
  Location(x::Int64, y::Int64) = new(x, y)
  Location() = Location(1, 1)
end

type Site
  owner :: Int64
  strength :: Int64
  production :: Int64
  Site(owner::Int64, strength::Int64, production::Int64) = new(owner, strength, production)
  Site() = Site(0, 0, 0)
end

type Move
  loc :: Location
  direction :: Int64
  Move(loc::Location, direction::Int64) = new(loc, direction)
  Move() = Move(0, Location())
end

type GameMap
  width :: Int64
  height :: Int64
  contents :: Vector{Vector{Site}}
  function GameMap(width::Int64, height::Int64, numberOfPlayers::Int64)
    contents = Vector{Vector{Site}}()
    for y in 1:height
      row = Vector{Site}()
      for x in 1:width
        push!(row, Site(0, 0, 0))
      end
      push!(contents, row)
    end
    new(width, height, contents)
  end
  GameMap(width::Int64, height::Int64) = GameMap(width, height, 0)
  GameMap() = GameMap(0, 0, 0)
end

isBounds(gm::GameMap, l::Location) = l.x > 0 && l.x <= gm.width && l.y > 0 && l.y <= gm.height

function getDeviation(gm::GameMap, l1::Location, l2::Location)
  dx = abs(l1.x - l2.x)
  dy = abs(l1.y - l2.y)
  if dx > gm.width / 2
    dx = gm.width - dx
  end
  if dy > gm.height / 2
    dy = gm.height - dy
  end
  (dx, dy)
end

function getDistance(gm::GameMap, l1::Location, l2::Location)
  dx, dy = getDeviation(gm, l1, l2)
  dx + dy
end

function getAngle(gm::GameMap, l1::Location, l2::Location)
  dx, dy = getDeviation(gm, l1, l2)
  atan2(dy, dx)
end

function getLocation(gm::GameMap, loc::Location, direction::Int64)
  l = deepcopy(loc)
  if direction != STILL
    if direction == NORTH
      if l.y == 1
        l.y = gm.height
      else
        l.y = l.y - 1
      end
    elseif direction == EAST
      if l.x == gm.width
        l.x = 1
      else
        l.x = l.x + 1
      end
    elseif direction == SOUTH
      if l.y == gm.height
        l.y = 1
      else
        l.y = l.y + 1
      end
    elseif direction == WEST
      if l.x == 1
        l.x = gm.width
      else
        l.x = l.x - 1
      end
    end
  end
  l
end

function getSite(gm::GameMap, l::Location, direction::Int64=STILL)
  l = getLocation(gm, l, direction)
  gm.contents[l.y][l.x]
end

File: airesources/PHP/hlt.php
<?php

define('STILL', 0);
define('NORTH', 1);
define('EAST', 2);
define('SOUTH', 3);
define('WEST', 4);

const DIRECTIONS = array(0, 1, 2, 3, 4);
const CARDINALS = array(1, 2, 3, 4);

define('ATTACK', 0);
define('STOP_ATTACK', 1);

class Location
{
    public $x;
    public $y;

    public function __construct($x = 0, $y = 0)
    {
        $this->x = $x;
        $this->y = $y;
    }
}

class Site
{
    public $owner;
    public $strength;
    public $production;

    public function __construct($owner = 0, $strength = 0, $production = 0)
    {
        $this->owner = $owner;
        $this->strength = $strength;
        $this->production = $production;
    }
}

class Move
{
    public $loc;
    public $direction;

    public function __construct(Location $loc, $direction = STILL)
    {
        $this->loc = $loc;
        $this->direction = $direction;
    }
}

class GameMap
{
    public $width;
    public $height;
    public $contents;

    public function __construct($width = 0, $height = 0, $numberOfPlayers = 0)
    {
        $this->width = $width;
        $this->height = $height;
        $this->contents = [];

        for ($y = 0; $y < $this->height; ++$y) {
            $row = [];
            for ($x = 0; $x < $this->width; ++$x) {
                $row[] = new Site(0, 0, 0);
            }
            $this->contents[] = $row;
        }
    }

    public function inBounds(Location $l)
    {
        return $l->x >= 0 && $l->x < $this->width && $l->y >= 0 && $l->y < $this->height;
    }

    public function getDistance(Location $l1, Location $l2)
    {
        $dx = abs($l1->x - $l2->x);
        $dy = abs($l1->y - $l2->y);
        if ($dx > $this->width / 2) {
            $dx = $this->width - $dx;
        }
        if ($dy > $this->height / 2) {
            $dy = $this->height - $dy;
        }
        return $dx + $dy;
    }

    public function getAngle(Location $l1, Location $l2)
    {
        $dx = $l2->x - $l1->x;
        $dy = $l2->y - $l1->y;

        if ($dx > $this->width - $dx) {
            $dx -= $this->width;
        } elseif (-$dx > $this->width + $dx) {
            $dx += $this->width;
        }

        if ($dy > $this->height - $dy) {
            $dy -= $this->height;
        } elseif (-$dy > $this->height + $dy) {
            $dy += $this->height;
        }
        return atan2($dy, $dx);
    }

    public function getLocation(Location $loc, $direction)
    {
        $l = clone $loc;
        if ($direction !== STILL) {
            if ($direction === NORTH) {
                if ($l->y === 0) {
                    $l->y = $this->height - 1;
                } else {
                    $l->y -= 1;
                }
            } elseif ($direction === EAST) {
                if ($l->x === $this->width - 1) {
                    $l->x = 0;
                } else {
                    $l->x += 1;
                }
            } elseif ($direction === SOUTH) {
                if ($l->y === $this->height - 1) {
                    $l->y = 0;
                } else {
                    $l->y += 1;
                }
            } elseif ($direction === WEST) {
                if ($l->x === 0) {
                    $l->x = $this->width - 1;
                } else {
                    $l->x -= 1;
                }
            }
        }
        return $l;
    }

    public function getSite(Location $l, $direction = STILL)
    {
        $l = $this->getLocation($l, $direction);
        return $this->contents[$l->y][$l->x];
    }
}
